File: granola_share/classify.py
"""Decide which class a note belongs to.

Order: the user's own Granola folder/title tags (rules) win; otherwise ask a local
Ollama model with a strict JSON schema; otherwise Unsorted.
"""

from __future__ import annotations

import json
import re
from typing import Callable

import httpx

from .config import UNSORTED, ClassDef, Config
from .granola import Meeting
from .store import Classification

MAX_NOTE_CHARS = 12000  # ~3k tokens; fits num_ctx below with room for the prompt


def _norm(s: str) -> str:
    return re.sub(r"[^a-z0-9]+", " ", (s or "").lower()).strip()


def classify_by_rules(m: Meeting, classes: list[ClassDef]) -> Classification | None:
    folder = _norm(m.folder)
    title = _norm(m.title)
    for c in classes:
        names = [c.name] + list(c.aliases)
        for n in names:
            nn = _norm(n)
            if not nn:
                continue
            if folder and (folder == nn or nn in folder.split(" / ") or nn in folder):
                return Classification(c.name, 0.95, "folder", lecture_title=m.title)
    for c in classes:
        for n in [c.name] + list(c.aliases):
            nn = _norm(n)
            if nn and len(nn) >= 3 and re.search(rf"\b{re.escape(nn)}\b", title):
                return Classification(c.name, 0.85, "rules", lecture_title=m.title)
    return None


def _schema(class_names: list[str]) -> dict:
    return {
        "type": "object",
        "properties": {
            "class_name": {"type": "string", "enum": class_names + [UNSORTED]},
            "confidence": {"type": "number"},
            "lecture_title": {"type": "string"},
            "topics": {"type": "array", "items": {"type": "string"}},
        },
        "required": ["class_name", "confidence", "lecture_title", "topics"],
    }


def build_prompt(m: Meeting, classes: list[ClassDef]) -> str:
    lines = ["You file lecture notes into the right class. Classes:"]
    for c in classes:
        extra = f" (also called: {', '.join(c.aliases)})" if c.aliases else ""
        desc = f" — {c.description}" if c.description else ""
        lines.append(f"- {c.name}{extra}{desc}")
    lines += [
        f"- {UNSORTED} — use this when the note is not a lecture for any class above, or you are unsure.",
        "",
        "Return JSON with: class_name (one of the classes above), confidence (0 to 1),",
        "lecture_title (a short specific title for this lecture), topics (3-6 short topic tags).",
        "",
        f"Title: {m.title}",
        f"Date: {m.date}",
        f"Granola folder: {m.folder or '(none)'}",
        f"Attendees: {', '.join(m.attendees) or '(none)'}",
        "",
        "Notes:",
        (m.notes_markdown or m.transcript or "")[:MAX_NOTE_CHARS],
    ]
    return "\n".join(lines)


def ollama_chat(cfg: Config, prompt: str, schema: dict) -> str:
    body = {
        "model": cfg.ollama_model,
        "messages": [{"role": "user", "content": prompt}],
        "stream": False,
        "format": schema,
        "think": False,
        "options": {"temperature": 0, "num_ctx": 8192},
    }
    r = httpx.post(f"{cfg.ollama_host.rstrip('/')}/api/chat", json=body, timeout=180)
    r.raise_for_status()
    return r.json()["message"]["content"]


def classify_with_ollama(m: Meeting, classes: list[ClassDef], cfg: Config,
                         chat: Callable[[Config, str, dict], str] = ollama_chat) -> Classification | None:
    if not classes:
        return None
    names = [c.name for c in classes]
    try:
        content = chat(cfg, build_prompt(m, classes), _schema(names))
        data = json.loads(content)
    except Exception:
        return None
    name = data.get("class_name")
    if name not in names + [UNSORTED]:
        return None
    try:
        conf = float(data.get("confidence", 0))
    except (TypeError, ValueError):
        conf = 0.0
    conf = max(0.0, min(1.0, conf))
    topics = [str(t) for t in data.get("topics", [])][:8]
    title = str(data.get("lecture_title") or m.title)
    if name == UNSORTED or conf < cfg.min_confidence:
        return Classification(UNSORTED, conf, "ollama", lecture_title=title, topics=topics)
    return Classification(name, conf, "ollama", lecture_title=title, topics=topics)


def classify(m: Meeting, cfg: Config, chat: Callable[[Config, str, dict], str] | None = None) -> Classification:
    c = classify_by_rules(m, cfg.classes)
    if c:
        return c
    if cfg.ollama_enabled and cfg.classes:
        c = classify_with_ollama(m, cfg.classes, cfg, chat or ollama_chat)
        if c:
            return c
    return Classification(UNSORTED, 0.0, "none", lecture_title=m.title)

File: engine/src/StudyStash.Core/Classify.cs
using System.Globalization;
using System.Text.Json;
using System.Text.Json.Nodes;
using System.Text.RegularExpressions;

namespace StudyStash.Core;

/// <summary>(cfg, prompt, JSON schema) → the model's answer: the sorting call.</summary>
public delegate Task<string> SortChatFn(Config cfg, string prompt, JsonObject schema);

/// <summary>
/// Decide which class a note belongs in. The student's own Granola folder or title wins; otherwise a local
/// Ollama model answers against a strict JSON schema; otherwise Unsorted.
/// </summary>
public static partial class Classify
{
    public const int MaxNoteChars = 12000; // ~3k tokens; fits the context below with room for the prompt

    [GeneratedRegex("[^a-z0-9]+")]
    private static partial Regex NotWordChars();

    internal static string Norm(string? s) => Py.Strip(NotWordChars().Replace((s ?? "").ToLowerInvariant(), " "));

    public static Classification? ByRules(Meeting m, IReadOnlyList<ClassDef> classes)
    {
        string folder = Norm(m.Folder);
        string title = Norm(m.Title);
        foreach (var c in classes)
        {
            foreach (string n in c.Aliases.Prepend(c.Name))
            {
                string nn = Norm(n);
                if (nn.Length == 0) continue;
                if (folder.Length > 0 && (folder == nn || folder.Contains(nn, StringComparison.Ordinal)))
                    return new Classification(c.Name, 0.95, "folder", m.Title);
            }
        }
        foreach (var c in classes)
        {
            foreach (string n in c.Aliases.Prepend(c.Name))
            {
                string nn = Norm(n);
                if (nn.Length >= 3 && Regex.IsMatch(title, $@"\b{Regex.Escape(nn)}\b"))
                    return new Classification(c.Name, 0.85, "rules", m.Title);
            }
        }
        return null;
    }

    public static JsonObject Schema(IEnumerable<string> classNames) => new()
    {
        ["type"] = "object",
        ["properties"] = new JsonObject
        {
            ["class_name"] = new JsonObject
            {
                ["type"] = "string",
                ["enum"] = new JsonArray(classNames.Append(Configs.Unsorted).Select(n => (JsonNode?)n).ToArray()),
            },
            ["confidence"] = new JsonObject { ["type"] = "number" },
            ["lecture_title"] = new JsonObject { ["type"] = "string" },
            ["topics"] = new JsonObject { ["type"] = "array", ["items"] = new JsonObject { ["type"] = "string" } },
        },
        ["required"] = new JsonArray("class_name", "confidence", "lecture_title", "topics"),
    };

    public static string BuildPrompt(Meeting m, IReadOnlyList<ClassDef> classes)
    {
        var lines = new List<string> { "You file lecture notes into the right class. Classes:" };
        foreach (var c in classes)
        {
            string extra = c.Aliases.Count > 0 ? $" (also called: {string.Join(", ", c.Aliases)})" : "";
            string desc = c.Description.Length > 0 ? $" \u2014 {c.Description}" : "";
            lines.Add($"- {c.Name}{extra}{desc}");
        }
        string notes = m.NotesMarkdown.Length > 0 ? m.NotesMarkdown : m.Transcript;
        lines.AddRange([
            $"- {Configs.Unsorted} \u2014 use this when the note is not a lecture for any class above, or you are unsure.",
            "",
            "Return JSON with: class_name (one of the classes above), confidence (0 to 1),",
            "lecture_title (a short specific title for this lecture), topics (3-6 short topic tags).",
            "",
            $"Title: {m.Title}",
            $"Date: {m.Date}",
            $"Granola folder: {(m.Folder.Length > 0 ? m.Folder : "(none)")}",
            $"Attendees: {(m.Attendees.Count > 0 ? string.Join(", ", m.Attendees) : "(none)")}",
            "",
            "Notes:",
            Py.Head(notes, MaxNoteChars),
        ]);
        return string.Join("\n", lines);
    }

    public static async Task<int> SortContextAsync(Config cfg)
    {
        // Same model as the summarizer: ask for the same context, or Ollama reloads it between the two calls.
        if (cfg.SummaryEnabled && cfg.OllamaModel == cfg.EffectiveSummaryModel)
            return await Summarize.ContextSizeAsync(cfg, cfg.OllamaModel);
        return 8192;
    }

    public static async Task<string> OllamaChatAsync(Config cfg, string prompt, JsonObject schema)
    {
        var body = new JsonObject
        {
            ["model"] = cfg.OllamaModel,
            ["messages"] = new JsonArray(new JsonObject { ["role"] = "user", ["content"] = prompt }),
            ["stream"] = false,
            ["format"] = schema.DeepClone(),
            ["think"] = false,
            // The answer is a small JSON object: a cap stops a model that loops from running on for minutes.
            ["options"] = new JsonObject { ["temperature"] = 0, ["num_ctx"] = await SortContextAsync(cfg), ["num_predict"] = 512 },
        };
        var data = await Ollama.PostAsync(cfg.OllamaHost, "/api/chat", body, TimeSpan.FromSeconds(300));
        if (Py.AsString(data?["done_reason"]) == "length")
            throw new RunawayOutputException("the sorting answer ran past 512 tokens"); // WithOllama then falls back
        return Py.AsString(data?["message"]?["content"]) ?? throw new InvalidDataException("Ollama sent no message");
    }

    public static async Task<Classification?> WithOllamaAsync(Meeting m, IReadOnlyList<ClassDef> classes, Config cfg,
        SortChatFn? chat = null)
    {
        if (classes.Count == 0) return null;
        var names = classes.Select(c => c.Name).ToList();
        JsonObject data;
        try
        {
            string content = await (chat ?? OllamaChatAsync)(cfg, BuildPrompt(m, classes), Schema(names));
            if (JsonNode.Parse(content) is not JsonObject o) return null;
            data = o;
        }
        catch (Exception)
        {
            return null; // Ollama down, a timeout, or an answer that isn't JSON: the note goes to Unsorted
        }
        string? name = Py.AsString(data["class_name"]);
        if (name is null || !(names.Contains(name) || name == Configs.Unsorted)) return null;
        double conf = Math.Clamp(Confidence(data["confidence"]), 0.0, 1.0);
        var topics = (data["topics"] as JsonArray ?? new JsonArray()).Select(Py.Str).Take(8).ToList();
        string title = Py.Truthy(data["lecture_title"]) ? Py.Str(data["lecture_title"]) : m.Title;
        if (name == Configs.Unsorted || conf < cfg.MinConfidence)
            return new Classification(Configs.Unsorted, conf, "ollama", title, topics);
        return new Classification(name, conf, "ollama", title, topics);
    }

    /// <summary>float(data.get("confidence", 0)), and 0 when that fails.</summary>
    static double Confidence(JsonNode? v)
    {
        switch (v)
        {
            case null:
                return 0;
            case JsonValue n when n.GetValueKind() == JsonValueKind.Number:
                return Py.NumberValue(n);
            case JsonValue b when b.GetValueKind() is JsonValueKind.True or JsonValueKind.False:
                return b.GetValueKind() == JsonValueKind.True ? 1 : 0;
            case JsonValue s when s.GetValueKind() == JsonValueKind.String:
                return double.TryParse(Py.Strip(s.GetValue<string>()), NumberStyles.Float, CultureInfo.InvariantCulture, out double d)
                    && !double.IsNaN(d) ? d : 0;
            default:
                return 0;
        }
    }

    public static async Task<Classification> ClassifyAsync(Meeting m, Config cfg, SortChatFn? chat = null)
    {
        var c = ByRules(m, cfg.Classes);
        if (c is not null) return c;
        if (cfg.OllamaEnabled && cfg.Classes.Count > 0)
        {
            c = await WithOllamaAsync(m, cfg.Classes, cfg, chat);
            if (c is not null) return c;
        }
        return new Classification(Configs.Unsorted, 0.0, "none", m.Title);
    }
}
